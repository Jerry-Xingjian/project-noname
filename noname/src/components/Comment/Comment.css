.comment .avatar-image {
  width: 32px;
  height: 32px;
}

.comment .mention-link {
  color: #0D8CFF;
}

.comment .commentInput{
  height: 60px;
  width: 90% !important;
}

.comment .commentInput__input{
  height: 60px;
  max-width: 100% !important;
  border-radius: 5px;
  /* border-color: transparent !important; */
  padding: 5px 10px !important;
  resize: none;
  overflow: scroll !important;
}

.comment .commentInput__highlighter {
  height: 60px;
  padding: 5px 10px !important;
}

.comment .mentionText {
  color: #6c757d;
  z-index: 1;
  position: relative;
}

.comment .mention-suggestion .avatar-image {
  width: 24px;
  height: 24px;
  border-radius: 50%;
  margin-right: 10px;
}

.comment .action-icon {
  width: 16px;
  height: 16px;
  transition: all 1s linear;
}

.comment .action-icon:hover {
  cursor: pointer;
}

.comment .action-icon:hover path{
  cursor: pointer;
  stroke: #CF250D
}

.comment .action-icon.heart.active path{
  fill: #CF250D;
  stroke: #CF250D;
}
