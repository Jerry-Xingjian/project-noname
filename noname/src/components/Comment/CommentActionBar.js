/* eslint-disable max-len */
/* eslint-disable no-underscore-dangle */
import React, { useState, useEffect } from 'react';
import { nFormatter } from '../../utils/formatters';
import { likePost, unlikePost } from '../../utils/api/post';

function CommentActionBar(props) {
  const { postInfo: post, userInfo } = props;
  const [heartClicked, setHeartClicked] = useState(false);
  const [postInfo, setPostInfo] = useState(() => ({
    ...post,
    like_count_str: '0',
  }));

  const ifUserLikedPost = () => post.likeInfo.filter((i) => `${i}` === `${userInfo._id}`).length;

  function fetchPostData() {
    if (!post.postUserInfo) {
      post.postUserInfo = userInfo;
    }
    setHeartClicked(ifUserLikedPost());
    post.like_count_str = nFormatter(post.likeInfo.length, 1);
    setPostInfo(post);
  }

  useEffect(() => {
    fetchPostData();
  }, [post]);

  const clickHeartHandler = async () => {
    setHeartClicked(!heartClicked);

    if (heartClicked) {
      // Unlike the post
      const res = await unlikePost(postInfo._id, userInfo._id);
      const editedPost = res.data.response;

      // Reset the post to update the like count
      const { postUserInfo } = postInfo;

      setPostInfo({
        ...editedPost,
        postUserInfo,
        like_count_str: nFormatter(editedPost.likeInfo.length, 1),
      });
    } else {
      // Like a post
      const res = await likePost(postInfo._id, userInfo._id);
      const editedPost = res.data.response;

      // Reset the post to update the like count
      const { postUserInfo } = postInfo;
      setPostInfo({
        ...editedPost,
        postUserInfo,
        like_count_str: nFormatter(editedPost.likeInfo.length, 1),
      });
    }
  };

  return (
    <div className="d-flex justify-content-between px-4 py-2 border-top border-bottom">
      {/* <svg className="action-icon" width="28" height="28" viewBox="0 0 28 29" fill="none" xmlns="http://www.w3.org/2000/svg">
        <path d="M9.08324 7.45L18.4894 4.11397C22.7105 2.61689 25.0039 5.0688 23.6079 9.56006L20.4725 19.5681C18.3675 26.2991 14.9108 26.2991 12.8058 19.5681L11.8752 16.5975L9.08324 15.6073C2.75708 13.3676 2.75708 9.70152 9.08324 7.45Z" stroke="#292D32" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" />
        <path d="M12.0857 16.0908L16.052 11.8589" stroke="#292D32" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" />
      </svg> */}
      <div className="d-flex align-items-center">
        <svg onClick={clickHeartHandler} className={`action-icon heart me-2 ${heartClicked ? 'active' : ''}`} width="26" height="25" viewBox="0 0 26 25" fill="none" xmlns="http://www.w3.org/2000/svg" role="presentation">
          <path d="M13.3679 21.6769C13.0138 21.8019 12.4304 21.8019 12.0763 21.6769C9.05542 20.6457 2.30542 16.3436 2.30542 9.05192C2.30542 5.83317 4.89917 3.229 8.09709 3.229C9.99292 3.229 11.67 4.14567 12.7221 5.56234C13.7742 4.14567 15.4617 3.229 17.3471 3.229C20.545 3.229 23.1388 5.83317 23.1388 9.05192C23.1388 16.3436 16.3888 20.6457 13.3679 21.6769Z" stroke="#292D32" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" />
        </svg>
        <p className="m-0">{postInfo.like_count_str}</p>
      </div>
    </div>
  );
}

export default CommentActionBar;
