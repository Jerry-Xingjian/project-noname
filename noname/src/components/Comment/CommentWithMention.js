import React from 'react';
import { Link } from 'react-router-dom';
import parse, { domToReact } from 'html-react-parser';

// options for html-react-parser
// Docmentation: https://github.com/remarkablemark/html-react-parser
const options = {
  replace: ({ attribs, name, children }) => {
    if (!attribs) {
      return null;
    }

    if (name === 'a') {
      const id = attribs.href;
      return <Link to={`/profile/${id}`} className="text-decoration-none">{ domToReact(children, options) }</Link>;
    }

    if (name === 'span') {
      return <span className="text-primary">{ domToReact(children, options) }</span>;
    }

    return null;
  },
};

function CommentWithMention(props) {
  const { contentObj: { plainTextValue, mentions } } = props;

  // Convert mentions to html
  const updatedMention = mentions.reduce((acc, mention) => {
    const { id, display } = mention;
    const html = `<a href=${id}><span>${display}</span></a>`;
    const newAcc = acc.replace(display, html);
    return newAcc;
  }, plainTextValue);

  return (
    <div>
      { parse(updatedMention, options) }
    </div>
  );
}

export default CommentWithMention;
