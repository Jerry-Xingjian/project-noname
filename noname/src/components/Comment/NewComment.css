.new-comment-bar .form-width {
    width: 90% !important;
}

.new-comment-bar .commentInput{
    height: 60px;
    width: 90% !important;
}

.new-comment-bar .commentInput__input{
    height: 60px;
    max-width: 100% !important;
    border-radius: 5px;
    border-color: transparent !important;
    padding: 5px 10px !important;
    resize: none;
    overflow: scroll !important;
}

.new-comment-bar .commentInput__highlighter {
    height: 60px;
    padding: 5px 10px !important;
}

.new-comment-bar .mentionText {
    color: #6c757d;
    z-index: 1;
    position: relative;
}

.new-comment-bar .mention-suggestion .avatar-image {
    width: 24px;
    height: 24px;
    border-radius: 50%;
    margin-right: 10px;
}