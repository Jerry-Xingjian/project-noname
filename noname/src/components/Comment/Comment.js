/* eslint-disable react/jsx-no-bind */
/* eslint-disable no-underscore-dangle */
/* eslint-disable max-len */
// CSS
import './Comment.css';
// React Component
import React, { useState, useEffect, useRef } from 'react';
import { MentionsInput, Mention } from 'react-mentions';
// Custom Component
import CommentWithMention from './CommentWithMention';
// API and helper functions
import data from '../../utils/api/index';
import { getUserByUserId } from '../../utils/api/user';
import { deleteComment, editComment } from '../../utils/api/comment';
import { timeAgoFormatter } from '../../utils/formatters';

function Comment(props) {
  const {
    commentInfo,
    postInfo,
    userInfo,
    setCommentsState,
    showConfirm,
    showAlert,
  } = props;
  const postId = postInfo._id;
  const [commentUser, setCommentUser] = useState({});
  const [commentCreated, setCommentCreated] = useState('1d');
  const [showEditInput, setShowEditInput] = useState(false);
  const [showActionIcons, setShowActionIcons] = useState(false);
  const [comment, setComment] = useState('');
  const commentObj = useRef({});

  useEffect(() => {
    async function getCommentUser() {
      const res = await getUserByUserId(commentInfo.belongUserId);
      setCommentUser(res.data.data);
      setCommentCreated(timeAgoFormatter(commentInfo.createdAt));
    }
    getCommentUser();
    setShowActionIcons(commentInfo.belongUserId === userInfo._id);
  }, []);

  const clickDeleteHandler = async () => {
    showConfirm(
      {
        show: true,
        title: 'Delete Comment',
        message: 'Are you sure to delete this comment?',
        confirmBtn: {
          text: 'Delete',
          variant: 'outline-secondary',
        },
        cancelBtn: {
          text: 'Cancel',
          variant: 'danger',
        },
      },
      async () => {
        const updatedPost = await deleteComment(postId, commentInfo._id);
        // Sort again after delete
        updatedPost.data.response.comments.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
        setCommentsState(updatedPost.data.response.comments);
      },
    );
  };

  const clickEditHandler = () => setShowEditInput(true);

  // Set the comment content value with mention whne comment change
  const handleCommentChange = async (e, _, newPlainTextValue, mentions) => {
    // Handle the changes in the textArea
    setComment(e.target.value);
    commentObj.current = {
      plainTextValue: newPlainTextValue,
      mentions,
    };
  };

  function fetchUsers(query, callback) {
    if (!query) return;
    fetch(`${data.rootTemp}/search/users?username=${query}`, {
      json: true,
    })
      .then((res) => res.json())
      // Transform the users to what react-mentions expects
      .then((res) => res.data.map((user) => ({ ...user, display: `@${user.username}`, id: user._id })))
      .then(callback);
  }

  // Function for a customeized mention dropdown list
  const renderSuggestion = (entry) => (
    <div className="d-flex px-1 py-1 mx-1 my-1 align-items-center mention-suggestion rounded">
      <img className="avatar-image me-3" src={entry.profilePicture} alt="Avatar" />
      <span>{entry.username}</span>
    </div>
  );

  const handleEditedComment = async (event) => {
    event.preventDefault();
    setShowEditInput(false);
    // Alert if the comment is empty
    if (comment === '' || comment.trim() === '') {
      showAlert('Comment content cannot be empty, please write a comment. ✍️');
      return;
    }

    // Clear comment box
    setComment('');

    // Init new comment object
    const postData = {
      belongPostId: postId,
      belongUserId: userInfo._id,
      content: commentObj.current,
    };

    // Make API call to add comment, then update comments state
    const commentId = commentInfo._id;
    const res = await editComment(postId, commentId, postData);
    const newComments = res.data.response.comments.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    setCommentsState(newComments);
  };

  return (
    <div>
      {
        showEditInput ? (
          <div className="d-flex comment mb-2 justify-content-between">
            <div className="d-flex w-100">
              <img className="avatar-image me-3" src={commentUser.profilePicture} alt={commentUser.username} />
              <div className="w-100">
                <p className="mb-1 fw-light">
                  <span className="fw-bold me-2">{commentUser.username}</span>
                </p>
                <form className="d-flex justify-content-between w-100" onSubmit={handleEditedComment}>
                  <MentionsInput
                    id="floatingTextarea"
                    className="commentInput w-100"
                    data-testid="comment-textarea"
                    name="comment"
                    value={comment}
                    onChange={handleCommentChange}
                    placeholder={commentInfo.content.plainTextValue}
                  >
                    <Mention
                      trigger="@"
                      data={fetchUsers}
                      renderSuggestion={renderSuggestion}
                      className="mentionText"
                    />
                  </MentionsInput>
                  <button data-testid="new-comment-post" type="submit" className="btn btn-link text-decoration-none">Update</button>
                </form>
              </div>
            </div>
          </div>
        ) : (
          <div className="d-flex comment mb-2 justify-content-between">
            <div className="d-flex">
              <img className="avatar-image me-3" src={commentUser.profilePicture} alt={commentUser.username} />
              <div>
                <p className="mb-1 fw-light">
                  <span className="fw-bold me-2">{commentUser.username}</span>
                </p>
                <CommentWithMention contentObj={commentInfo.content} />
                <p className="mb-1 text-black-50">{commentCreated}</p>
              </div>
            </div>
            <div className={showActionIcons ? 'd-flex align-items-center' : 'd-flex align-items-center d-none'}>
              {/* Edit Icon */}
              <svg onClick={clickEditHandler} className="mx-2 action-icon" width="16" height="16" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                <path d="M13.26 3.59997L5.04997 12.29C4.73997 12.62 4.43997 13.27 4.37997 13.72L4.00997 16.96C3.87997 18.13 4.71997 18.93 5.87997 18.73L9.09997 18.18C9.54997 18.1 10.18 17.77 10.49 17.43L18.7 8.73997C20.12 7.23997 20.76 5.52997 18.55 3.43997C16.35 1.36997 14.68 2.09997 13.26 3.59997Z" stroke="#121212" strokeWidth="1.5" strokeMiterlimit="10" strokeLinecap="round" strokeLinejoin="round" />
                <path d="M11.89 5.04999C12.32 7.80999 14.56 9.91999 17.34 10.2" stroke="#121212" strokeWidth="1.5" strokeMiterlimit="10" strokeLinecap="round" strokeLinejoin="round" />
                <path d="M3 22H21" stroke="#121212" strokeWidth="1.5" strokeMiterlimit="10" strokeLinecap="round" strokeLinejoin="round" />
              </svg>
              {/* Delete Icon */}
              <svg onClick={clickDeleteHandler} className="mx-2 action-icon" width="16" height="16" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                <path d="M21 5.98C17.67 5.65 14.32 5.48 10.98 5.48C9 5.48 7.02 5.58 5.04 5.78L3 5.98" stroke="#121212" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" />
                <path d="M8.5 4.97L8.72 3.66C8.88 2.71 9 2 10.69 2H13.31C15 2 15.13 2.75 15.28 3.67L15.5 4.97" stroke="#121212" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" />
                <path d="M18.85 9.14001L18.2 19.21C18.09 20.78 18 22 15.21 22H8.79C6 22 5.91 20.78 5.8 19.21L5.15 9.14001" stroke="#121212" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" />
                <path d="M10.33 16.5H13.66" stroke="#121212" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" />
                <path d="M9.5 12.5H14.5" stroke="#121212" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" />
              </svg>
            </div>
          </div>
        )
      }
    </div>
  );
}

export default Comment;
