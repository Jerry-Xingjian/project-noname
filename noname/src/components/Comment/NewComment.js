/* eslint-disable no-undef */
/* eslint-disable react/jsx-no-bind */
/* eslint-disable no-underscore-dangle */
import React, { useRef, useState } from 'react';
import { MentionsInput, Mention } from 'react-mentions';
import './NewComment.css';
import data from '../../utils/api/index';
import { addComment } from '../../utils/api/comment';

function NewComment(props) {
  const [comment, setComment] = useState('');
  const commentObj = useRef({});
  const {
    userInfo,
    postInfo,
    setCommentsState,
    showAlert,
  } = props;

  // Async function to query user from database
  // Should limit the number of users returned to 5
  function fetchUsers(query, callback) {
    if (!query) return;
    fetch(`${data.rootTemp}/search/users?username=${query}`, {
      json: true,
    })
      .then((res) => res.json())
      // Transform the users to what react-mentions expects
      .then((res) => res.data.map((user) => ({ ...user, display: `@${user.username}`, id: user._id })))
      .then(callback);
  }

  // Set the comment content value with mention whne comment change
  const handleCommentChange = async (e, _, newPlainTextValue, mentions) => {
    // Handle the changes in the textArea
    setComment(e.target.value);
    commentObj.current = {
      plainTextValue: newPlainTextValue,
      mentions,
    };
  };

  // Post new comment
  const handleNewComment = async (event) => {
    event.preventDefault();
    // Alert if the comment is empty
    if (comment === '' || comment.trim() === '') {
      showAlert('Comment content cannot be empty, please write a comment. ✍️');
      return;
    }
    // Clear comment box
    setComment('');
    // Init new comment object
    const postData = {
      belongPostId: postInfo._id,
      belongUserId: userInfo._id,
      content: commentObj.current,
    };

    // Make API call to add comment, then update comments state
    const res = await addComment(postData);
    let newComments = res.data.response.comments;
    // Sort the comments by createdAt
    newComments = newComments.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    setCommentsState(newComments);
  };

  // Function for a customeized mention dropdown list
  const renderSuggestion = (entry) => (
    <div className="d-flex px-1 py-1 mx-1 my-1 align-items-center mention-suggestion rounded">
      <img className="avatar-image me-3" src={entry.profilePicture} alt="Avatar" />
      <span>{entry.username}</span>
    </div>
  );

  return (
    <div className="new-comment-bar px-4 py-2 d-flex align-items-center">
      <img className="avatar-image me-3" src={userInfo.profilePicture} alt="Avatar" />
      <form className="d-flex form-width justify-content-between w-100" onSubmit={handleNewComment}>
        <MentionsInput
          id="floatingTextarea"
          className="commentInput w-100"
          data-testid="comment-textarea"
          name="comment"
          value={comment}
          onChange={handleCommentChange}
          placeholder="Leave a comment here"
        >
          <Mention
            trigger="@"
            // data={users}
            data={fetchUsers}
            renderSuggestion={renderSuggestion}
            className="mentionText"
          />
        </MentionsInput>
        <button data-testid="new-comment-post" type="submit" className="btn btn-link text-decoration-none">Post</button>
      </form>
    </div>
  );
}

export default NewComment;
