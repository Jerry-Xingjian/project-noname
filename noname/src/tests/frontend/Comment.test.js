import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Comment from '../../components/Comment/Comment';
import renderer from 'react-test-renderer';

const commentInfo = {
  belong_user: {
    post_id: '1',
    comment_id: "9b3166e5-860c-4215-9bfb-53cf6cc2b156",
    belong_post_id: '1',
    like_count: 81433,
    created_at: "2022-10-16T05:37:51.403Z",
    updated_at: "2022-10-17T17:23:13.877Z",
    belong_user_id: "30084f19-273f-4575-8882-482a7ce5a8e7",
    profile_picture: "http://loremflickr.com/640/480/fashion",
  },
  content: {
    plainTextValue: 'great comment@dylan',
    mentions: [12, "@"],
  },
}

describe("Snapshot test", () => {
  test('Comment component matches snapshot', () => {
    const component = renderer.create(
      <Comment commentInfo={commentInfo} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Component test", () => {
  test('certain texts exists in the page', () => {
    render(<Comment commentInfo={commentInfo}/>);
    const text = screen.getByText(/great comment@dylan/i);
    expect(text).toBeInTheDocument();
  })
});