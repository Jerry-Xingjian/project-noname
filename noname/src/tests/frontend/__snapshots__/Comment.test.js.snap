// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Comment component matches snapshot 1`] = `
<div
  className="d-flex comment mb-2 justify-content-between"
>
  <div
    className="d-flex"
  >
    <img
      className="avatar-image me-3"
    />
    <div>
      <p
        className="mb-1 fw-light"
      >
        <span
          className="fw-bold me-2"
        />
      </p>
      <div>
        great comment@dylan
      </div>
      <p
        className="mb-1 text-black-50"
      >
        1d
      </p>
    </div>
  </div>
  <div
    className="d-flex align-items-center"
  >
    <svg
      className="mx-2"
      fill="none"
      height="16"
      onClick={[Function]}
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 5.98C17.67 5.65 14.32 5.48 10.98 5.48C9 5.48 7.02 5.58 5.04 5.78L3 5.98"
        stroke="#121212"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="1.5"
      />
      <path
        d="M8.5 4.97L8.72 3.66C8.88 2.71 9 2 10.69 2H13.31C15 2 15.13 2.75 15.28 3.67L15.5 4.97"
        stroke="#121212"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="1.5"
      />
      <path
        d="M18.85 9.14001L18.2 19.21C18.09 20.78 18 22 15.21 22H8.79C6 22 5.91 20.78 5.8 19.21L5.15 9.14001"
        stroke="#121212"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="1.5"
      />
      <path
        d="M10.33 16.5H13.66"
        stroke="#121212"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="1.5"
      />
      <path
        d="M9.5 12.5H14.5"
        stroke="#121212"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="1.5"
      />
    </svg>
  </div>
</div>
`;
